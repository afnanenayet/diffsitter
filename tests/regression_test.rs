#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use libdiffsitter::{
        diff::compute_edit_script, generate_ast_vector_data, input_processing::TreeSitterProcessor,
        parse::GrammarConfig,
    };
    use std::path::PathBuf;
    use test_case::test_case;

    /// Get paths to input files for tests
    fn get_test_paths(test_type: &str, test_name: &str, ext: &str) -> (PathBuf, PathBuf) {
        let test_data_root = PathBuf::from(format!("./test_data/{test_type}/{test_name}"));
        let path_a = test_data_root.join(format!("a.{ext}"));
        let path_b = test_data_root.join(format!("b.{ext}"));
        assert!(
            path_a.exists(),
            "test data path {} does not exist",
            path_a.to_str().unwrap()
        );
        assert!(
            path_b.exists(),
            "test data path {} does not exist",
            path_b.to_str().unwrap()
        );

        (path_a, path_b)
    }

    #[test_case("short", "rust", "rs", true, true)]
    #[test_case("short", "python", "py", true, true)]
    #[test_case("short", "go", "go", true, true)]
    #[test_case("medium", "rust", "rs", true, false)]
    #[test_case("medium", "rust", "rs", false, false)]
    #[test_case("medium", "cpp", "cpp", true, true)]
    #[test_case("medium", "cpp", "cpp", false, true)]
    fn diff_hunks_snapshot(
        test_type: &str,
        name: &str,
        ext: &str,
        split_graphemes: bool,
        strip_whitespace: bool,
    ) {
        let (path_a, path_b) = get_test_paths(test_type, name, ext);
        let config = GrammarConfig::default();
        let ast_data_a = generate_ast_vector_data(path_a, None, &config).unwrap();
        let ast_data_b = generate_ast_vector_data(path_b, None, &config).unwrap();

        let processor = TreeSitterProcessor {
            split_graphemes,
            strip_whitespace,
            ..Default::default()
        };

        let diff_vec_a = processor.process(&ast_data_a.tree, &ast_data_a.text);
        let diff_vec_b = processor.process(&ast_data_b.tree, &ast_data_b.text);
        let diff_hunks = compute_edit_script(&diff_vec_a, &diff_vec_b).unwrap();

        // We have to set the snapshot name manually, otherwise there appear to be threading issues
        // and we end up with more snapshot files than there are tests, which cause
        // nondeterministic errors.
        let snapshot_name = format!("{test_type}_{name}_split_graphemes_{split_graphemes}_strip_whitespace_{strip_whitespace}");
        assert_debug_snapshot!(snapshot_name, diff_hunks);
    }
}
