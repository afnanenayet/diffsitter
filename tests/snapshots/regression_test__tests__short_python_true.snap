---
source: tests/regression_test.rs
expression: diff_hunks
---
RichHunks(
    [
        New(
            Hunk(
                [
                    Line {
                        line_index: 1,
                        entries: [
                            Entry {
                                reference: {Node string_start (1, 4) - (1, 7)},
                                text: "\"",
                                start_position: Point {
                                    row: 1,
                                    column: 4,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 5,
                                },
                                kind_id: 104,
                            },
                            Entry {
                                reference: {Node string_start (1, 4) - (1, 7)},
                                text: "\"",
                                start_position: Point {
                                    row: 1,
                                    column: 5,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 6,
                                },
                                kind_id: 104,
                            },
                            Entry {
                                reference: {Node string_start (1, 4) - (1, 7)},
                                text: "\"",
                                start_position: Point {
                                    row: 1,
                                    column: 6,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 7,
                                },
                                kind_id: 104,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 1,
                                    column: 7,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 8,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "H",
                                start_position: Point {
                                    row: 1,
                                    column: 8,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 9,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "e",
                                start_position: Point {
                                    row: 1,
                                    column: 9,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 10,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "l",
                                start_position: Point {
                                    row: 1,
                                    column: 10,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 11,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "l",
                                start_position: Point {
                                    row: 1,
                                    column: 11,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 12,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "o",
                                start_position: Point {
                                    row: 1,
                                    column: 12,
                                },
                                end_position: Point {
                                    row: 1,
                                    column: 13,
                                },
                                kind_id: 231,
                            },
                        ],
                    },
                    Line {
                        line_index: 2,
                        entries: [
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 2,
                                    column: 0,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 1,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 2,
                                    column: 1,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 2,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 2,
                                    column: 2,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 3,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 2,
                                    column: 3,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 4,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "w",
                                start_position: Point {
                                    row: 2,
                                    column: 4,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 5,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "o",
                                start_position: Point {
                                    row: 2,
                                    column: 5,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 6,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "r",
                                start_position: Point {
                                    row: 2,
                                    column: 6,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 7,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "l",
                                start_position: Point {
                                    row: 2,
                                    column: 7,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 8,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: "d",
                                start_position: Point {
                                    row: 2,
                                    column: 8,
                                },
                                end_position: Point {
                                    row: 2,
                                    column: 9,
                                },
                                kind_id: 231,
                            },
                        ],
                    },
                    Line {
                        line_index: 3,
                        entries: [
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 3,
                                    column: 0,
                                },
                                end_position: Point {
                                    row: 3,
                                    column: 1,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 3,
                                    column: 1,
                                },
                                end_position: Point {
                                    row: 3,
                                    column: 2,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 3,
                                    column: 2,
                                },
                                end_position: Point {
                                    row: 3,
                                    column: 3,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_content (1, 7) - (3, 4)},
                                text: " ",
                                start_position: Point {
                                    row: 3,
                                    column: 3,
                                },
                                end_position: Point {
                                    row: 3,
                                    column: 4,
                                },
                                kind_id: 231,
                            },
                            Entry {
                                reference: {Node string_end (3, 4) - (3, 7)},
                                text: "\"",
                                start_position: Point {
                                    row: 3,
                                    column: 4,
                                },
                                end_position: Point {
                                    row: 3,
                                    column: 5,
                                },
                                kind_id: 107,
                            },
                            Entry {
                                reference: {Node string_end (3, 4) - (3, 7)},
                                text: "\"",
                                start_position: Point {
                                    row: 3,
                                    column: 5,
                                },
                                end_position: Point {
                                    row: 3,
                                    column: 6,
                                },
                                kind_id: 107,
                            },
                            Entry {
                                reference: {Node string_end (3, 4) - (3, 7)},
                                text: "\"",
                                start_position: Point {
                                    row: 3,
                                    column: 6,
                                },
                                end_position: Point {
                                    row: 3,
                                    column: 7,
                                },
                                kind_id: 107,
                            },
                        ],
                    },
                ],
            ),
        ),
        New(
            Hunk(
                [
                    Line {
                        line_index: 6,
                        entries: [
                            Entry {
                                reference: {Node identifier (6, 4) - (6, 5)},
                                text: "x",
                                start_position: Point {
                                    row: 6,
                                    column: 4,
                                },
                                end_position: Point {
                                    row: 6,
                                    column: 5,
                                },
                                kind_id: 1,
                            },
                            Entry {
                                reference: {Node = (6, 10) - (6, 11)},
                                text: "=",
                                start_position: Point {
                                    row: 6,
                                    column: 10,
                                },
                                end_position: Point {
                                    row: 6,
                                    column: 11,
                                },
                                kind_id: 44,
                            },
                            Entry {
                                reference: {Node integer (6, 12) - (6, 13)},
                                text: "1",
                                start_position: Point {
                                    row: 6,
                                    column: 12,
                                },
                                end_position: Point {
                                    row: 6,
                                    column: 13,
                                },
                                kind_id: 93,
                            },
                        ],
                    },
                ],
            ),
        ),
    ],
)
